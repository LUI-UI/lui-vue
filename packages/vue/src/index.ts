export { default as LuiAvatar } from './components/Avatar/LuiAvatar.vue'
export { default as LuiBadge } from './components/Badge/LuiBadge.vue'
export { default as LuiButton } from './components/Button/LuiButton.vue'
export { default as LuiCheckbox } from './components/Checkbox/LuiCheckbox.vue'
export { default as LuiMenuDropdown } from './components/MenuDropdown/LuiMenuDropdown.vue'
export { default as LuiMenuItem } from './components/MenuItem/LuiMenuItem.vue'
export { default as LuiInput } from './components/Input/LuiInput.vue'
export { default as LuiRadio } from './components/Radio/LuiRadio.vue'
export { default as LuiSelect } from './components/Select/LuiSelect.vue'
export { default as LuiMultiSelect } from './components/MultiSelect/LuiMultiSelect.vue'
export { default as LuiOption } from './components/Option/LuiOption.vue'
export { default as LuiSwitch } from './components/Switch/LuiSwitch.vue'
export { default as LuiTextarea } from './components/Textarea/LuiTextarea.vue'
export { default as LuiModal } from './components/Modal/LuiModal.vue'
export { default as LuiTabGroup } from './components/Tabs/LuiTabGroup.vue'
export { default as LuiTabButtons } from './components/Tabs/LuiTabButtons.vue'
export { default as LuiTabButton } from './components/Tabs/LuiTabButton.vue'
export { default as LuiTabPanels } from './components/Tabs/LuiTabPanels.vue'
export { default as LuiTabPanel } from './components/Tabs/LuiTabPanel.vue'
export { default as LuiTag } from './components/Tag/LuiTag.vue'
export { default as LuiPopover } from './components/Popover/LuiPopover.vue'
export { default as LuiSidebar } from './components/Sidebar/LuiSidebar.vue'
