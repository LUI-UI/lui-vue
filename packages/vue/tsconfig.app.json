{
  "extends": "@vue/tsconfig/tsconfig.web.json",
  "compilerOptions": {
    "composite": true,
    "baseUrl": ".",
    "paths": {
      "@*": ["./src/*"]
    }
  },
  "include": ["env.d.ts", "src/**/*", "src/**/*.vue", "tailwind.config.cjs"],
  "exclude": ["src/**/__tests__/*"]
}
