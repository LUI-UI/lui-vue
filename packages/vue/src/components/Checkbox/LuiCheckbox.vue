<script lang="ts">
export default {
  name: 'LuiCheckbox',
  inheritAttrs: false,
}
</script>

<script setup lang="ts">
import type { PropType } from 'vue'
import { computed, toRefs, useAttrs } from 'vue'
import { useGlobalDescriptionClasses } from '../../composables'
import type {
  CheckableModelValue,
  Color,
  Description,
  Rounded,
  Size,
  State,
  Value,
} from '../../globals/types'
import { useCheckboxClasses } from './composables/index'

type Indeterminate = false | true

const props = defineProps({
  size: {
    type: String as PropType<Size>,
    default: 'md',
  },
  rounded: {
    type: [Boolean, String] as PropType<Rounded>,
    default: false,
  },
  state: {
    type: [String, Boolean, null] as PropType<State>,
    default: null,
  },
  description: {
    type: [String, null] as PropType<Description>,
    default: null,
  },
  indeterminate: {
    type: Boolean as PropType<Indeterminate>,
    default: false,
  },
  modelValue: {
    type: [Array, Boolean, String] as PropType<CheckableModelValue>,
    required: false,
  },
  value: {
    type: [String, Number, Boolean, null, undefined] as PropType<Value>,
    required: false,
  },
  trueValue: {
    type: [String, Number, Boolean, null, undefined] as PropType<Value>,
    required: false,
  },
  falseValue: {
    type: [String, Number, Boolean, null, undefined] as PropType<Value>,
    required: false,
  },
  color: {
    type: String as PropType<Color>,
    default: 'primary',
  },
})
const emit = defineEmits(['update:modelValue', 'change'])
const attrs = useAttrs()
const { inputClasses, spanClasses, iconClasses } = useCheckboxClasses(
  toRefs(props),
)
const { descriptionClasses } = useGlobalDescriptionClasses(
  toRefs(props),
  attrs,
)

const iconSize = computed(() =>
  // 12 - 16 - 20 - 24 - 28
  props.size === 'xs'
    ? {
        checkbox: '12',
        indeterminate: { width: '10', stroke: '1.5', viewBox: '0 0 10 2' },
      }
    : props.size === 'sm'
      ? {
          checkbox: '16',
          indeterminate: { width: '12', stroke: '1.75', viewBox: '0 0 12 2' },
        }
      : props.size === 'md'
        ? {
            checkbox: '20',
            indeterminate: { width: '16', stroke: '2', viewBox: '0 0 16 2' },
          }
        : props.size === 'lg'
          ? {
              checkbox: '24',
              indeterminate: { width: '18', stroke: '2', viewBox: '0 0 18 2' },
            }
          : {
              checkbox: '28',
              indeterminate: { width: '22', stroke: '2', viewBox: '0 0 22 2' },
            },
)

function handleChange(event: any) {
  const target = event.target as HTMLInputElement
  const isChecked = target.checked
  if (props.modelValue && Array.isArray(props.modelValue) && props.value) {
    const newValue = [...props.modelValue]
    if (isChecked)
      newValue.push(props.value)
    else newValue.splice(newValue.indexOf(props.value), 1)

    handleEmits(newValue, event)
  }
  else if (props.trueValue && props.falseValue) {
    const currentValue = isChecked ? props.trueValue : props.falseValue
    handleEmits(currentValue, event)
  }
  else {
    handleEmits(isChecked, event)
  }
}
const isChecked = computed(() => {
  if (!props.modelValue)
    return attrs.checked as boolean
  if (Array.isArray(props.modelValue) && props.value)
    return props.modelValue.includes(props.value)
  else if (props.trueValue && props.falseValue)
    return props.modelValue === props.trueValue
  else return props.modelValue as boolean
})
function handleEmits(value: CheckableModelValue, event: any) {
  emit('change', event)
  emit('update:modelValue', value)
}
</script>

<template>
  <div class="lui-checkbox inline-block leading-3 group">
    <div class="relative inline-flex">
      <input
        class="checkbox"
        type="checkbox"
        v-bind="$attrs"
        :class="inputClasses"
        :checked="isChecked"
        :value="value"
        @change="handleChange"
      >
      <span :class="spanClasses" />
      <svg
        v-if="!indeterminate"
        :class="iconClasses"
        :width="iconSize.checkbox"
        :height="iconSize.checkbox"
        viewBox="0 0 12 12"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.64286 3.42857L4.63393 8.57143L2.35714 6.23376"
          class="stroke-white dark:stroke-secondary-300"
          stroke-width="1.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
      <svg
        v-if="indeterminate && size === 'xs'"
        width="10"
        height="2"
        viewBox="0 0 10 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="stroke-white dark:stroke-secondary-800"
          d="M1 1H9"
          stroke-width="1.5"
          stroke-linecap="round"
        />
      </svg>
      <svg
        v-if="indeterminate && size === 'sm'"
        width="12"
        height="2"
        viewBox="0 0 12 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="stroke-white dark:stroke-secondary-800"
          d="M1 1H11"
          stroke-width="1.75"
          stroke-linecap="round"
        />
      </svg>
      <svg
        v-if="indeterminate && size === 'md'"
        width="16"
        height="2"
        viewBox="0 0 16 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="stroke-white dark:stroke-secondary-800"
          d="M1 1H15"
          stroke-width="2"
          stroke-linecap="round"
        />
      </svg>
      <svg
        v-if="indeterminate && size === 'lg'"
        width="18"
        height="2"
        viewBox="0 0 18 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="stroke-white dark:stroke-secondary-800"
          d="M1 1H17"
          stroke-width="2"
          stroke-linecap="round"
        />
      </svg>
      <svg
        v-if="indeterminate && size === 'xl'"
        width="22"
        height="2"
        viewBox="0 0 22 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="stroke-white dark:stroke-secondary-800"
          d="M1 1H21"
          stroke-width="2"
          stroke-linecap="round"
        />
      </svg>
    </div>
    <p v-if="description !== null" :class="descriptionClasses">
      {{ description }}
    </p>
  </div>
</template>
