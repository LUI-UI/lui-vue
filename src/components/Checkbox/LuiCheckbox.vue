<script lang="ts">
export default {
  name: 'LuiCheckbox',
  inheritAttrs: false
}
</script>
<script setup lang="ts">
import type { PropType } from 'vue'
import type { Rounded, Size, State, Description, CheckableModelValue } from '@/globals/types'
import { toRefs, computed, useAttrs } from 'vue'
import { useCheckboxClasses } from './composables/index'
import { useGlobalDescriptionClasses } from '../../composables/index'
import { useGlobalCheckbox } from '../../composables/index'

type Indeterminate = false | true

const props = defineProps({
  size: {
    type: String as PropType<Size>,
    default: 'md'
  },
  rounded: {
    type: [Boolean, String] as PropType<Rounded>,
    default: false,
    validator(value: Rounded) {
      return [true, false, 'full'].includes(value)
    }
  },
  state: {
    type: [String, Boolean, null] as PropType<State>,
    default: null
  },
  description: {
    type: [String, null] as PropType<Description>,
    default: null
  },
  indeterminate: {
    type: Boolean as PropType<Indeterminate>,
    default: false
  },
  value: {
    type: [String, Boolean],
    default: false
  },
  modelValue: {
    type: [Array, Boolean, undefined] as PropType<CheckableModelValue>,
    default: undefined
  }
  // indeterminate
})
const attrs = useAttrs()
const { inputClasses, spanClasses, iconClasses } = useCheckboxClasses(toRefs(props))
const { descriptionClasses } = useGlobalDescriptionClasses(toRefs(props), attrs)
const { handleVModel, isInputChecked } = useGlobalCheckbox(props, attrs)
const emit = defineEmits(['update:modelValue'])

function handleChange(e: any) {
  emit('update:modelValue', handleVModel(e))
}

const iconSize = computed(() =>
  // 12 - 16 - 20 - 24 - 28
  props.size === 'xs'
    ? {
        checkbox: '12',
        indeterminate: { width: '10', stroke: '1.5', viewBox: '0 0 10 2' }
      }
    : props.size === 'sm'
    ? {
        checkbox: '16',
        indeterminate: { width: '12', stroke: '1.75', viewBox: '0 0 12 2' }
      }
    : props.size === 'md'
    ? {
        checkbox: '20',
        indeterminate: { width: '16', stroke: '2', viewBox: '0 0 16 2' }
      }
    : props.size === 'lg'
    ? {
        checkbox: '24',
        indeterminate: { width: '18', stroke: '2', viewBox: '0 0 18 2' }
      }
    : {
        checkbox: '28',
        indeterminate: { width: '22', stroke: '2', viewBox: '0 0 22 2' }
      }
)
</script>
<template>
  <div class="inline-block leading-3">
    <div class="relative inline-flex">
      <input
        type="checkbox"
        :checked="isInputChecked"
        :value="value"
        @change="handleChange"
        :class="inputClasses"
        v-bind="$attrs"
      />
      <span :class="spanClasses"> </span>
      <svg
        v-if="!indeterminate"
        :class="iconClasses"
        :width="iconSize.checkbox"
        :height="iconSize.checkbox"
        viewBox="0 0 12 12"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.64286 3.42857L4.63393 8.57143L2.35714 6.23376"
          stroke="currentColor"
          stroke-width="1.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
      <!-- <svg
        v-if="indeterminate"
        :width="iconSize.indeterminate.width"
        height="2"
        viewBox="0 0 16 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.999977 1H15"
          stroke="white"
          :stroke-width="iconSize.indeterminate.stroke"
          stroke-linecap="round"
        />
      </svg> -->
      <svg
        v-if="indeterminate && size === 'xs'"
        width="10"
        height="2"
        viewBox="0 0 10 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path d="M1 1H9" stroke="white" stroke-width="1.5" stroke-linecap="round" />
      </svg>
      <svg
        v-if="indeterminate && size === 'sm'"
        width="12"
        height="2"
        viewBox="0 0 12 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path d="M1 1H11" stroke="white" stroke-width="1.75" stroke-linecap="round" />
      </svg>
      <svg
        v-if="indeterminate && size === 'md'"
        width="16"
        height="2"
        viewBox="0 0 16 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path d="M1 1H15" stroke="white" stroke-width="2" stroke-linecap="round" />
      </svg>
      <svg
        v-if="indeterminate && size === 'lg'"
        width="18"
        height="2"
        viewBox="0 0 18 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path d="M1 1H17" stroke="white" stroke-width="2" stroke-linecap="round" />
      </svg>
      <svg
        v-if="indeterminate && size === 'xl'"
        width="22"
        height="2"
        viewBox="0 0 22 2"
        :class="iconClasses"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path d="M1 1H21" stroke="white" stroke-width="2" stroke-linecap="round" />
      </svg>
    </div>
    <p v-if="description !== null" :class="descriptionClasses">
      {{ description }}
    </p>
  </div>
</template>
