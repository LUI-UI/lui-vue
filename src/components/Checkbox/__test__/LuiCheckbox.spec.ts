import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import type { DOMWrapper, VueWrapper } from '@vue/test-utils'
import { mount } from '@vue/test-utils'
import { ref } from 'vue'
import LuiCheckbox from '../LuiCheckbox.vue'

describe('Default LuiCheckbox', () => {
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiCheckbox)
  })
  afterEach(() => {
    wrapper.unmount()
  })
  it('renders with default props', async () => {
    const defaultProps = {
      rounded: false,
      state: null,
      indeterminate: false,
      size: 'md',
      description: null,
      modelValue: undefined,
    }
    await wrapper.vm.$nextTick
    expect(wrapper.props()).toEqual(defaultProps)
    expect(wrapper.exists()).toBe(true)
    expect(wrapper.html).toMatchSnapshot()
  })

  // it('renders indeterminate checkbox when indeterminate prop is given', async () => {
  //   await wrapper.setProps({ indeterminate: true })
  //   // the classic styled svg is not rendering when indeterminate prop is true
  //   expect(
  //     wrapper.find('[data-testid="lui-checkbox-classic-svg"]').exists(),
  //   ).toBe(false)
  //   // the indeterminate styled svg is rendering when indeterminate prop is true
  //   expect(
  //     wrapper.find('[data-testid="lui-checkbox-indeterminate-svg"]').exists(),
  //   ).toBe(true)
  //   // there is only one indeterminate styled svg is rendering
  //   expect(
  //     wrapper.findAll('[data-testid="lui-checkbox-indeterminate-svg"]').length,
  //   ).toBe(1)
  // })
  it('renders as checked when checked prop is given', async () => {
    await wrapper.setProps({ checked: true })
    const inputEl = wrapper.find(
      '[data-testid=\'lui-checkbox-input\']',
    ) as DOMWrapper<HTMLInputElement>
    // input rendered as checked beforehand the users interaction
    expect(inputEl.element.checked).toBe(true)
  })
  it('emits change event whenever an inputs value changed', () => {
    const inputEl = wrapper.find('[data-testid=\'lui-checkbox-input\']') as any
    expect(wrapper.emitted('change')).toBeFalsy()
    inputEl.setChecked(true)
    expect(wrapper.emitted('change')).toBeTruthy()
  })

  it('defines disabled attribute when disabled prop is given', async () => {
    await wrapper.setProps({ disabled: true })

    const inputEl = wrapper.find('[data-testid="lui-checkbox-input"]') as any
    // check if attributes has come to inputEl
    expect(inputEl.attributes('disabled')).toBeDefined()
  })

  it('should properly inherit attributes', () => {
    const id = 'myId'

    const wrapper = mount(LuiCheckbox, {
      attrs: {
        id,
      },
    })
    // Check if the custom attribute exists on the input element
    expect(
      wrapper.find('[data-testid="lui-checkbox-input"]').attributes('id'),
    ).toBe(id)
  })
})

describe('LuiCheckbox with v-model', () => {
  it('works properly with boolean v-model', async () => {
    const testVal = ref(false)
    const wrapper = mount(LuiCheckbox, {
      props: { 'onUpdate:modelValue': (e: any) => (testVal.value = e) },
    })

    const inputEl = wrapper.find(
      '[data-testid=\'lui-checkbox-input\']',
    ) as DOMWrapper<HTMLInputElement>

    // initally input element is not checked
    expect(inputEl.element.checked).toBe(false)
    // check if v-model working properly when input's value changed by testing emitted values
    inputEl.setValue(true)

    await wrapper.vm.$nextTick()
    expect(testVal.value).toBe(true)
    wrapper.unmount()
  })

  it('works properly with array v-model', async () => {
    const checkedNames = ref([])
    const firstCheckbox = mount(LuiCheckbox, {
      props: {
        'onUpdate:modelValue': (e: any) => (checkedNames.value = e),
        'modelValue': checkedNames.value,
        'value': 'John',
      },
    })
    const secondCheckbox = mount(LuiCheckbox, {
      props: {
        'onUpdate:modelValue': (e: any) => (checkedNames.value = e),
        'modelValue': checkedNames.value,
        'value': 'Jack',
      },
    })

    const firstInput = firstCheckbox.find(
      '[data-testid="lui-checkbox-input"]',
    ) as any
    const secondInput = secondCheckbox.find(
      '[data-testid="lui-checkbox-input"]',
    ) as any

    firstInput.setChecked(true)
    expect(checkedNames.value).toEqual(['John'])
    secondInput.setChecked(true)
    expect(checkedNames.value).toEqual(['John', 'Jack'])
    firstInput.setChecked(false)
    expect(checkedNames.value).toEqual(['Jack'])

    // unmount wrappers
    secondCheckbox.unmount()
    firstCheckbox.unmount()
  })
  it('works properly with custom values', async () => {
    const customTrueValue = 'yes'
    const customFalseValue = 'no'
    const customValue = ref('') // Start with the false value

    const wrapper = mount(LuiCheckbox, {
      props: {
        'modelValue': customValue.value,
        'onUpdate:modelValue': (e: any) => (customValue.value = e),
      },
      attrs: {
        'true-value': customTrueValue,
        'false-value': customFalseValue,
      },
    })
    const inputEl = wrapper.find('[data-testid="lui-checkbox-input"]') as any
    // tests for custom true value
    inputEl.setChecked(true)
    // Wait for Vue to update
    await wrapper.vm.$nextTick()
    expect(customValue.value).toEqual(customTrueValue)
    // tests for custom false value
    inputEl.setChecked(false)
    await wrapper.vm.$nextTick()
    expect(inputEl.element.checked).toBe(false)
    expect(customValue.value).toEqual(customFalseValue)
    wrapper.unmount()
  })
})

describe('LuiCheckbox description', () => {
  const descriptionExample = 'A description'
  let wrapper: VueWrapper<any>
  beforeEach(() => {
    wrapper = mount(LuiCheckbox)
  })
  afterEach(() => {
    wrapper.unmount()
  })
  it('renders description properly from description prop', async () => {
    await wrapper.setProps({ description: descriptionExample })
    expect(
      wrapper.find('[data-testid="lui-checkbox-description"]').exists(),
    ).toBe(true)
    expect(
      wrapper.find('[data-testid="lui-checkbox-description"]').text(),
    ).toEqual(descriptionExample)
  })
  // it('renders description properly when state prop is null', async () => {
  //   await wrapper.setProps({ description: descriptionExample, state: null })

  //   expect(
  //     wrapper.find('[data-testid="lui-checkbox-description"]').classes(),
  //   ).toContain('text-secondary-600')
  // })
  // it('renders description properly when state prop is true', async () => {
  //   await wrapper.setProps({ description: descriptionExample, state: true })

  //   expect(
  //     wrapper.find('[data-testid="lui-checkbox-description"]').classes(),
  //   ).toContain('text-success-500')
  // })
  // it('renders description properly when state prop is warning', async () => {
  //   await wrapper.setProps({
  //     description: descriptionExample,
  //     state: 'warning',
  //   })
  //   expect(
  //     wrapper.find('[data-testid="lui-checkbox-description"]').classes(),
  //   ).toContain('text-warning-500')
  // })
  // it('renders description properly when state prop is false', async () => {
  //   await wrapper.setProps({ description: descriptionExample, state: false })

  //   expect(
  //     wrapper.find('[data-testid="lui-checkbox-description"]').classes(),
  //   ).toContain('text-danger-500')
  // })
})
