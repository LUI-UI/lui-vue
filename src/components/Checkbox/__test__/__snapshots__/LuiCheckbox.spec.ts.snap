// Vitest Snapshot v1

exports[`Default LuiCheckbox > renders with default props 1`] = `[Function]`;
